### ISSUE TITLE
Optionally break up long types on `->` when pretty-printing

### ISSUE THREAD

**Issue by byorgey** [swarm-game/swarm#1473]
**Describe the bug**
This example code:
```
def harvestbox : dir -> (cmd unit -> cmd unit) -> (cmd unit -> cmd unit) -> cmd unit = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end
```
is currently pretty-printed as
```
def harvestbox : dir -> (cmd unit -> cmd unit) -> (
    cmd unit -> cmd unit
  ) -> cmd unit = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end;
```
which looks weird.  I would expect/desire something more like the following, perhaps:
```
def harvestbox : 
  dir ->
  (cmd unit -> cmd unit) ->
  (cmd unit -> cmd unit) ->
  cmd unit
  = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end;
```
I think the issue is that right now, the only option the pretty-printer has for spreading long function types across multiple lines is to split out parenthesized parts.  But I think we instead need to pretty-print function types in such a way that the pretty-printer knows the right way to optionally break lines at `->`, which I think would mean that parenthesized types are rarely split across multiple lines like that.  But I could imagine it still being the right thing to do in a case like
```
def foo : 
  lots ->
  of ->
  parameters ->
  (
    a ->
    very ->
    long ->
    function ->
    parameter
  ) ->
  more ->
  result
```
So I don't think there's anything wrong with the way we handle parenthesized types per se.

**Comment by xsebek** [comment]
@byorgey I looked into this and the problem is that we want to basically get alternative layout for _lambda chain_.

How do I get that chain though? This current instance does not even know that it is a fixed type:
```Haskell
instance (PrettyPrec t) => PrettyPrec (TypeF t) where
instance (PrettyPrec (t (Fix t))) => PrettyPrec (Fix t) where
```

If I used `Type` instead, it would run into overlapping instances. 😖

I think moving the fix instance to a plain function and using it in the fixed type instances would be the most clean solution.

**Comment by xsebek** [comment]
OK, I see getting rid of Fixed instance would lead to duplicating code... so maybe we could add some class instance for both `Fix` and `UTerm` to please give us the lambda chain...

This recursive business is breaking my mind a bit and I got stuck, so any pointers would be appreciated. 😅

**Comment by byorgey** [comment]
Hmm, yes, I see what you mean... yes, some kind of type class for unchaining function arrows seems like our best bet.  So maybe something like this (untested):
```
class Unchainable t where
  unchain :: t -> [t]

instance Unchainable Type where
  unchain (a :->: ty) = a : unchain ty
  unchain ty = [ty]

instance Unchainable UType where
  unchain (UTyFun a ty) = a : unchain ty
  unchain ty = [ty]

instance (Unchainable t, PrettyPrec t) => PrettyPrec (TypeF t) where
  prettyPrec p (TyFunF ty1 ty2) = ... do something with  (ty1 : unchain ty2) ... 
  ... all other cases remain the same ...
```

**Comment by xsebek** [comment]
Btw. here is a quick script I came up with to see the different layouts from 10 to 100 columns in terminal:
```bash
t=$(mktemp -d)
for i in {10..100}; do cabal run swarm -O0 -- format --width $i example.sw > $t/$i.sw; done
for i in {10..100}; do if ! cmp -s $t/$i.sw $t/$(($i + 1)).sw; then echo $i; echo '```'; cat $t/$i.sw; echo '```';fi; done
```

<details>
<summary>Click to expand </summary>

21
```
def harvestbox:
  dir ->
  (
    cmd unit ->
    cmd unit
  ) ->
  (
    cmd unit ->
    cmd unit
  ) ->
  cmd unit 
  = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end
```
37
```
def harvestbox: dir ->
                (
                  cmd unit ->
                  cmd unit
                ) ->
                (
                  cmd unit ->
                  cmd unit
                ) ->
                cmd unit 
  = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end
```
40
```
def harvestbox: dir ->
                (
                  cmd unit -> cmd unit
                ) ->
                (
                  cmd unit -> cmd unit
                ) ->
                cmd unit 
  = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end
```
43
```
def harvestbox: dir ->
                (cmd unit -> cmd unit) ->
                (cmd unit -> cmd unit) ->
                cmd unit 
  = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end
```
83
```
def harvestbox: dir ->
                (cmd unit -> cmd unit) ->
                (cmd unit -> cmd unit) ->
                cmd unit = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end
```
100
```
def harvestbox: dir -> (cmd unit -> cmd unit) -> (cmd unit -> cmd unit) -> cmd unit 
  = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end
```

</details>

**Comment by byorgey** [comment]
Nice script!   I am not a big fan of the variants where we align the entire type to the column following the `:`, leaving a giant space to the left.  As a general principle I like the idea that the amount of indentation should never depend on the length of some identifier(s) in previous lines, but should always be some consistent, constant amount. The format with 37 columns looks particularly ridiculous --- wasting all that empty space to the left and then breaking up the type into a ton of lines in order to squish it into a narrow vertical space between the `:` and the margin.  Personally, when going from 100 to 83 I would like to see something like
```
def harvestbox: 
  dir ->
  (cmd unit -> cmd unit) ->
  (cmd unit -> cmd unit) ->
  cmd unit 
  = \d. \rep1. \rep2.
    doboxP harvest d rep1 rep2 notempty
end
```
In other words, as soon as the entire type is too long to fit on one line, then we stick newlines *before and after* it, indent the whole thing two spaces, and break on `->`.  How hard do you think that would that be to do?

**Comment by xsebek** [comment]
@byorgey not hard to do at all! What we need is a `flatAlt` on `def`.

It is a bit hard to refactor the code and not make it blow up. The same layout idea should also apply to let with slight modifications. 😅

**Comment by xsebek** [comment]
@byorgey this was a bit tricky, what do you think now @byorgey? 🙂 

150
```
def harvestbox: dir -> (cmd unit -> cmd unit) -> (cmd unit -> cmd unit) -> cmd unit = \d. \rep1. \rep2. doboxP harvest d rep1 rep2 notempty end
```
140
```
def harvestbox: dir -> (cmd unit -> cmd unit) -> (cmd unit -> cmd unit) -> cmd unit = \d. \rep1. \rep2. doboxP harvest d rep1 rep2 notempty
end
```
90
```
def harvestbox: dir -> (cmd unit -> cmd unit) -> (cmd unit -> cmd unit) -> cmd unit 
  = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end
```
30
```
def harvestbox: 
    dir ->
    (cmd unit -> cmd unit) ->
    (cmd unit -> cmd unit) ->
    cmd unit 
  = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end
```
10
```
def harvestbox: 
    dir ->
    (
      cmd unit ->
      cmd unit
    ) ->
    (
      cmd unit ->
      cmd unit
    ) ->
    cmd unit 
  = \d. \rep1. \rep2.
  doboxP harvest d rep1 rep2 notempty
end
```

**Comment by byorgey** [comment]
@xsebek that looks perfect!
